### PR TITLE
feat: overhaul

### PR DESCRIPTION
**Pull request by mert-kirpici**
This commit prepares previously private reusable worflows to be used
directly by the calling project while removing the 2nd order reusable
workflows for PR and release.

Apart from the main change above, it also:
- merges snap build and snap release workflow into a single one.
- move matrix strategy out of the lint-unit workflow

Signed-off-by: Mert Kırpıcı <mert.kirpici@canonical.com>

### CODE REVIEW

**Review by rgildein — COMMENTED** [review]
Overall it looks good to me, but I'm not sure if it's ok to remove lint, unit and functional tests from release workflows (charm and snap). What do you think @esunar? It should be fine, since we cannot merge anything directly to main or master branch, but maybe PR will be merged without passing those tests.

**Comment by mert-kirpici** [comment]
> Overall it looks good to me, but I'm not sure if it's ok to remove lint, unit and functional tests from release workflows (charm and snap). What do you think @esunar? It should be fine, since we cannot merge anything directly to main or master branch, but maybe PR will be merged without passing those tests.

I was thinking we reuse the check(lint-unit-func) workflow in the release workflow on the calling side(project) by adding workflow_call to `pr.yaml` to make it reusable and then adding this to the `release.yaml`:
```
jobs:
  tests:
    uses: ./.github/workflows/pr.yaml

  snap-release:
    uses: canonical/bootstack-actions/.github/workflows/snap-release.yaml@v2
    needs: tests
    ...
```

tested in private repo. seems to work fine.

**Comment by mert-kirpici** [comment]
> That's not nice (calling PR not in PR :) ), but it make sense.

we can rename it to `check.yaml` ? it is a check pipeline afterall

File: .github/workflows/func.yaml [hunk 1]
```diff
@@ -14,7 +14,7 @@ on:
           Tox version, which should be used, e.g. `<4`. If not specified, the
           latest version of tox will be installed.
         type: string
-        default: ""
+        default: "<4"
```
Comment (agileshaw) [inline .github/workflows/func.yaml:17]: We need to change the description if using `tox<4` as default. The current wording gives the wrong impression that "no user input == unspecified".

**Review by agileshaw — CHANGES_REQUESTED** [review]
I like the overall changes. Just one wording change suggested inline. 

About the on-going discussion, I think including tests in the release workflow is necessary because passing tests in PR doesn't guarantee an issue-free merge. The idea of putting lint-unit-func tests in a `check.yaml` file (or maybe naming it `ci-validation.yaml`) which then be used in both PR and release workflow sounds nice and clean.

**Review by rgildein — APPROVED** [review]
Let's try it. Thanks

**Review by esunar — APPROVED** [review]

**Review by agileshaw — APPROVED** [review]
Looks good, thanks

**Pull request merged**